src/libbson/src/bson/bson-json.c

```diff
@@ -978,13 +978,28 @@ _bson_json_read_string (bson_json_reader_t *reader, /* IN */
 {
    bson_json_read_state_t rs;
    bson_json_read_bson_state_t bs;
+   bool allow_null = true;
 
    BASIC_CB_PREAMBLE;
 
    rs = bson->read_state;
    bs = bson->bson_state;
 
-   if (!bson_utf8_validate ((const char *) val, vlen, true /*allow null*/)) {
+   /* Prohibit embedded NULL bytes in BSON regex pattern or options. */
+   if (rs == BSON_JSON_IN_BSON_TYPE_REGEX_VALUES &&
+       (bs == BSON_JSON_LF_REGULAR_EXPRESSION_PATTERN ||
+        bs == BSON_JSON_LF_REGULAR_EXPRESSION_OPTIONS)) {
+      /* Prohibit embedded NULL bytes for canonical extended regex: {
+       * $regularExpression: { pattern: "pattern", options: "options" } } */
+      allow_null = false;
+   } else if (rs == BSON_JSON_IN_BSON_TYPE &&
+              (bs == BSON_JSON_LF_REGEX || bs == BSON_JSON_LF_OPTIONS)) {
+      /* Prohibit embedded NULL bytes for legacy regex: { $regex: "pattern",
+       * $options: "options" } */
+      allow_null = false;
+   }
+
+   if (!bson_utf8_validate ((const char *) val, vlen, allow_null)) {
```

Comment:
I always get worried about very complex boolean logic dances. Could this be extracted into a function that checks a `bson_json_reader_bson_t ` object specifically to calculate the `allow_null` value? i.e.

```c
static bool
_bson_json_allow_embedded_nulls (bson_json_reader_t const *reader)
{
   const bson_json_read_state_t read_state = reader->bson.read_state;
   const bson_json_read_bson_state_t bson_state = reader->bson.bson_state;

   if (read_state == BSON_JSON_IN_BSON_TYPE_REGEX_VALUES) {
      if (bson_state == BSON_JSON_LF_REGULAR_EXPRESSION_PATTERN ||
          bson_state == BSON_JSON_LF_REGULAR_EXPRESSION_OPTIONS) {
         /* Prohibit embedded NULL bytes for canonical extended regex:
          * { $regularExpression: { pattern: "pattern", options: "options" } }
          */
         return false;
      }
   }

   if (read_state == BSON_JSON_IN_BSON_TYPE) {
      if (bson_state == BSON_JSON_LF_REGEX |
          bson_state == BSON_JSON_LF_OPTIONS) {
         /* Prohibit embedded NULL bytes for legacy regex:
          * { $regex: "pattern", $options: "options" } */
         return false;
      }
   }

   /* Embedded nulls are okay in any other context */
   return true;
}
```

And then `allow_null` can be `const`-initialized:

```c
const bool allow_null = _bson_json_allow_embedded_nulls(reader);
```

or even just omitted entirely:

```c
if (!bson_utf8_validate((const char*)val,
                        vlen, 
                        _bson_json_allow_embedded_nulls(reader)) {
```

This is mostly just a stylistic choice.

Reply:
I like it, and thank you for the code snippet. Done.